kroxylicious-systemtests/src/test/java/io/kroxylicious/systemtests/KroxyliciousST.java

```diff
@@ -166,9 +163,9 @@ private void scaleKroxylicious(String namespace, int replicas, int scaleTo) {
         assertThat(currentReplicas).withFailMessage("unexpected current scaled replicas").isEqualTo(scaleTo);
 
         LOGGER.atInfo().setMessage("Then the messages are consumed").log();
-        String result = KroxyliciousSteps.consumeMessages(namespace, topicName, bootstrap, expectedMessage, numberOfMessages, Duration.ofMinutes(2));
+        String result = KroxyliciousSteps.consumeMessages(namespace, topicName, bootstrap, MESSAGE, numberOfMessages, Duration.ofMinutes(2));
         LOGGER.atInfo().setMessage("Received: {}").addArgument(result).log();
-        assertThat(result).withFailMessage("expected message have not been received!").contains(expectedMessage);
+        assertThat(StringUtils.countMatches(result, MESSAGE)).withFailMessage("expected messages have not been received!").isEqualTo(numberOfMessages);
```

Comment:
This says to me that the `consumeMessages` abstraction is leaky. We shouldn't be reaching the number of messages assertion without this being `true`. 

I'm not sure what I think the correct shape should be but I'm starting to think passing `MESSAGE` into `consumeMessages` is wrong. I'm not sure we should even be passing `numberOfMessages` but I think that has more merit to it. `consumeMessages` should just return what ever is in the topic and shouldn't try to validate the messages it consumes its then up to tests to ensure they have the messages they expect. (The reason I think `numberOfMessages` is different is that its more like setting a batch size, and saying I want to read X messages from the topic)

Reply:
> I'm not sure what I think the correct shape should be but I'm starting to think passing MESSAGE into consumeMessages is wrong.

+1

> I'm not sure we should even be passing numberOfMessages but I think that has more merit to it. consumeMessages should just return what ever is in the topic and shouldn't try to validate the messages it consumes its then up to tests to ensure they have the messages they expect. 

Without numberOfMessages, consume would need to know how it should be prepared to wait.

Reply:
Why would it wait? If there is nothing on the topic it could just return empty string/ empty list.

I'm not actually sure what I think the best api around this. Waiting for x messages doesn't seem unreasonable but is it trying to push to much of the test level concerns down into the test infrastructure?

Reply:
I tried to explain it in my comment below. But I will resume it here as well.

Consumers are kubernetes jobs that could be run forever if we don't stop it, so we decided to stop it for 2 reasons:
1. Number of messages received are the expected
2. Timeout

Having the first option we can go faster in our tests, instead of always waiting for a timeout.

In our test cases we need to check again that the number of expected messages are correct because we don't know the status of the consumer.

Reply:
> Why would it wait? If there is nothing on the topic it could just return empty string/ empty list.

Fetching in Kafka always involves a wait.  It is baked into the protocol.   Maybe I miss your point.